From 1d1ab5254f38496b6abe9e687aa47782a5b91f66 Mon Sep 17 00:00:00 2001
From: Billy Tsai <billy_tsai@aspeedtech.com>
Date: Wed, 20 Oct 2021 18:47:49 +0800
Subject: [PATCH 3/3] drivers: peci: Add raw command for peci testing.

The raw command can test all of the peci instruction.

Signed-off-by: Billy Tsai <billy_tsai@aspeedtech.com>
Change-Id: I374629a87ef4ae4aa4b14f0dd7e336c7d1809f35
---
 drivers/peci/peci_shell.c                     | 39 +++++++++++++++++++
 .../ast1030_evb/demo/boards/ast1030_evb.conf  |  1 +
 2 files changed, 40 insertions(+)

diff --git a/drivers/peci/peci_shell.c b/drivers/peci/peci_shell.c
index 7cdef5d04c..357bac51e6 100644
--- a/drivers/peci/peci_shell.c
+++ b/drivers/peci/peci_shell.c
@@ -133,6 +133,43 @@ static int cmd_gettemp(const struct shell *shell, size_t argc, char **argv)
 	return 0;
 }
 
+static int cmd_raw(const struct shell *shell, size_t argc, char **argv)
+{
+	const struct device *dev;
+	struct peci_msg packet;
+	uint32_t index;
+	int ret, argv_index;
+	uint8_t peci_req_buf[64];
+	uint8_t peci_resp_buf[64];
+
+	dev = device_get_binding(argv[1]);
+	if (!dev) {
+		shell_error(shell, "peci device not found");
+		return -EINVAL;
+	}
+
+	packet.addr = strtoul(argv[2], NULL, 0);
+	packet.cmd_code = strtoul(argv[5], NULL, 16);
+	packet.tx_buffer.buf = peci_req_buf;
+	packet.tx_buffer.len = strtoul(argv[3], NULL, 0);
+	packet.rx_buffer.buf = peci_resp_buf;
+	packet.rx_buffer.len = strtoul(argv[4], NULL, 0);
+
+	for (argv_index = 6; argv_index < argc; argv_index++) {
+		peci_req_buf[argv_index - 6] = strtoul(argv[argv_index], NULL, 0);
+	}
+
+	ret = peci_transfer(dev, &packet);
+	if (ret) {
+		printk("Transfer failed %d\n", ret);
+		return ret;
+	}
+	for (index = 0; index < packet.rx_buffer.len; index++)
+		printk("%02x ", peci_resp_buf[index]);
+	printk("\n");
+	return 0;
+}
+
 SHELL_STATIC_SUBCMD_SET_CREATE(
 	peci_cmds,
 	SHELL_CMD_ARG(init, NULL, "<device> <kbps>", cmd_init, 3,
@@ -143,6 +180,8 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
 		      0),
 	SHELL_CMD_ARG(gettemp, NULL, "<device> <addr>", cmd_gettemp, 3,
 		      0),
+	SHELL_CMD_ARG(raw, NULL, "<device> <addr> <wr_len> <rd_len> <command(hex)> <...>", cmd_raw, 6,
+		      64),
 	SHELL_SUBCMD_SET_END);
 
 SHELL_CMD_REGISTER(peci, &peci_cmds, "PECI shell commands", NULL);
diff --git a/samples/boards/ast1030_evb/demo/boards/ast1030_evb.conf b/samples/boards/ast1030_evb/demo/boards/ast1030_evb.conf
index 486a59390a..1b7c7b0845 100644
--- a/samples/boards/ast1030_evb/demo/boards/ast1030_evb.conf
+++ b/samples/boards/ast1030_evb/demo/boards/ast1030_evb.conf
@@ -27,3 +27,4 @@ CONFIG_FLASH_PAGE_LAYOUT=y
 CONFIG_IMG_BLOCK_BUF_SIZE=4096
 CONFIG_USB_REQUEST_BUFFER_SIZE=4096
 CONFIG_USB_DFU_DEFAULT_POLLTIMEOUT=32
+CONFIG_SHELL_ARGC_MAX=70
-- 
2.25.1

